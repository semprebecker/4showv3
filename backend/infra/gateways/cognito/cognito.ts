import * as AWS from "aws-sdk";

import { getSecrets } from "../secretsmanager";

type Output = { cognitoId: string };

export class Cognito {
  protected cognitoConnection;
  constructor() {
    this.cognitoConnection = new AWS.CognitoIdentityServiceProvider({
      apiVersion: "2016-04-18",
    });
  }

  protected async setupConfig(): Promise<Output> {
    const secrets = await getSecrets();
    this.cognitoConnection.config.credentials!.accessKeyId = secrets.keyId;
    this.cognitoConnection.config.credentials!.secretAccessKey =
      secrets.accessKey;
    return { cognitoId: secrets.cognitoId };
  }
}
