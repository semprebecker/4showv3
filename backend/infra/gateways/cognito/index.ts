export * from './cognito'
export * from './user'
