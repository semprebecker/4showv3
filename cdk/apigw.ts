import { WebSocketApi, WebSocketStage } from "@aws-cdk/aws-apigatewayv2-alpha";
import { WebSocketLambdaIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha";
import { RemovalPolicy, Stack } from "aws-cdk-lib";
import { CfnAccount } from "aws-cdk-lib/aws-apigateway";
import { CfnModel, CfnRoute, CfnStage } from "aws-cdk-lib/aws-apigatewayv2";
import {
  ManagedPolicy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import { Function as LambdaFunction } from "aws-cdk-lib/aws-lambda";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import * as cognito from "aws-cdk-lib/aws-cognito";

import { MessageAction } from "../backend/types";
import { getCfnAuthorizer } from "./authorizer";
import { lambdaFunctions } from "./lambda";
import { getModels } from "./models";
import { envs } from "../backend/main/config";

export const stageName = envs.stage;

/*
 * Utility function for adding a route with a model.
 * This is not supported in the L2 constuct so we must drop to
 * L1 (Cfn) by casting the `WebSocketRoute` to `CfnRoute` (`AWS::ApiGatewayV2::Route`)
 */
const addRoute = (
  handler: LambdaFunction,
  action: MessageAction,
  model: CfnModel,
  webSocketApi: WebSocketApi
) => {
  const route = webSocketApi.addRoute(action, {
    integration: new WebSocketLambdaIntegration(
      `${action}Integration`,
      handler
    ),
  });
  const rt = route.node.defaultChild as CfnRoute;

  rt.modelSelectionExpression = "$request.body.action";
  rt.requestModels = { [action]: model.name };

  rt.addDependsOn(model);
};

/*
 * Again, we need to extend the `WebSocketStage` construct by casting it to `CfnStage` (`AWS::ApiGatewayV2::Stage`).
 */
const getStageAndLogs = (
  scope: Stack,
  webSocketApi: WebSocketApi
): WebSocketStage => {
  // Unlike RestApi, you don't get a default `prod` stage automatically.
  const stage = new WebSocketStage(scope, `${envs.stage.toUpperCase()}Stage`, {
    autoDeploy: true,
    stageName,
    webSocketApi,
  });
  // Manage the log group
  new LogGroup(scope, "ExecutionLogs", {
    logGroupName: `/aws/apigateway/${webSocketApi.apiId}/${stageName}`,
    removalPolicy: RemovalPolicy.DESTROY,
    retention: RetentionDays.ONE_WEEK,
  });
  const log = new LogGroup(scope, "AccessLogs", {
    removalPolicy: RemovalPolicy.DESTROY,
    retention: RetentionDays.ONE_WEEK,
  });
  const cfnStage = stage.node.defaultChild as CfnStage;
  cfnStage.accessLogSettings = {
    destinationArn: log.logGroupArn,
    format: `$context.identity.sourceIp - - [$context.requestTime] "$context.httpMethod $context.routeKey $context.protocol" $context.status $context.responseLength $context.requestId`,
  };
  cfnStage.defaultRouteSettings = {
    dataTraceEnabled: true,
    loggingLevel: "INFO",
    throttlingBurstLimit: 500,
    throttlingRateLimit: 1000,
  };

  /*
   * This role is automatically created by the RestApi construct but not by WebSocketApi.
   * CfnAccount isn't even available in the `aws-cdk-lib/aws-apigatewayv2` lib so we must import `aws-cdk-lib/aws-apigateway`
   * to create the CloudWatch role.
   */
  const cwRole = new Role(scope, "CWRole", {
    assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
    managedPolicies: [
      ManagedPolicy.fromAwsManagedPolicyName(
        "service-role/AmazonAPIGatewayPushToCloudWatchLogs"
      ),
    ],
  });

  const account = new CfnAccount(scope, "Account", {
    cloudWatchRoleArn: cwRole.roleArn,
  });

  webSocketApi.node.addDependency(account);
  return stage;
};

export const getApiGateway = (
  scope: Stack,
  fns: lambdaFunctions
): [WebSocketApi, WebSocketStage] => {
  const webSocketApi = new WebSocketApi(
    scope,
    `WebSocketApi4show${envs.stage.toUpperCase()}`,
    {
      connectRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "OnConnectIntegration",
          fns["setup-connection/onConnect"]
        ),
      },
      disconnectRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "OnDisconnectIntegration",
          fns["setup-connection/onDisconnect"]
        ),
      },
    }
  );

  const {
    getStateModel,
    addUser,
    updateUser,
    getUser,
    addAddressUser,
    updateAddressUser,
    getAddressesUser,
    listProfilesAdmin,
    listPermissionsAdmin,
    manageProfilePermissionsAdmin,
    listProfilePermissionsAdmin,
    addAccountAdmin,
    updateAccountAdmin,
    deleteAccountAdmin,
    listAccountsAdmin,
    countPagesAccountsAdmin,
    manageAccountLinkedUsersAdmin,
    listAccountLinkedUsersAdmin,
    countPagesUsersAdmin,
  } = getModels(scope, webSocketApi);

  getCfnAuthorizer(scope, webSocketApi, fns["setup-connection/authorizer"]);

  addRoute(
    fns["setup-connection/getState"],
    MessageAction.GET_STATE,
    getStateModel,
    webSocketApi
  );

  addRoute(fns["user/addUser"], MessageAction.ADD_USER, addUser, webSocketApi);

  addRoute(
    fns["user/updateUser"],
    MessageAction.UPDATE_USER,
    updateUser,
    webSocketApi
  );

  addRoute(fns["user/getUser"], MessageAction.GET_USER, getUser, webSocketApi);

  addRoute(
    fns["user/addAddress"],
    MessageAction.ADD_ADDRESS_USER,
    addAddressUser,
    webSocketApi
  );

  addRoute(
    fns["user/updateAddress"],
    MessageAction.UPDATE_ADDRESS_USER,
    updateAddressUser,
    webSocketApi
  );

  addRoute(
    fns["user/getAddresses"],
    MessageAction.GET_ADDRESSES_USER,
    getAddressesUser,
    webSocketApi
  );

  addRoute(
    fns["admin/listProfiles"],
    MessageAction.LIST_PROFILES,
    listProfilesAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listPermissions"],
    MessageAction.LIST_PERMISSIONS,
    listPermissionsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/manageProfilePermissions"],
    MessageAction.MANAGE_PROFILE_PERMISSIONS,
    manageProfilePermissionsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listProfilePermissions"],
    MessageAction.LIST_PROFILE_PERMISSIONS,
    listProfilePermissionsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/addAccount"],
    MessageAction.ADD_ACCOUNT,
    addAccountAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/updateAccount"],
    MessageAction.UPDATE_ACCOUNT,
    updateAccountAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/deleteAccount"],
    MessageAction.DELETE_ACCOUNT,
    deleteAccountAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listAccounts"],
    MessageAction.LIST_ACCOUNTS,
    listAccountsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/countPagesAccounts"],
    MessageAction.COUNT_PAGES_ACCOUNTS,
    countPagesAccountsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/manageAccountLinkedUsers"],
    MessageAction.MANAGE_ACCOUNT_LINKED_USERS,
    manageAccountLinkedUsersAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listAccountLinkedUsers"],
    MessageAction.LIST_ACCOUNT_LINKED_USERS,
    listAccountLinkedUsersAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/countPagesUsers"],
    MessageAction.COUNT_PAGES_USERS,
    countPagesUsersAdmin,
    webSocketApi
  );

  // These functions need to be able to manage websocket connections.
  [
    fns["setup-connection/getState"],
    fns["user/addUser"],
    fns["user/updateUser"],
    fns["user/getUser"],
    fns["user/addAddress"],
    fns["user/updateAddress"],
    fns["user/getAddresses"],
    fns["admin/listProfiles"],
    fns["admin/listPermissions"],
    fns["admin/manageProfilePermissions"],
    fns["admin/listProfilePermissions"],
    fns["admin/addAccount"],
    fns["admin/updateAccount"],
    fns["admin/deleteAccount"],
    fns["admin/listAccounts"],
    fns["admin/countPagesAccounts"],
    fns["admin/manageAccountLinkedUsers"],
    fns["admin/listAccountLinkedUsers"],
    fns["admin/countPagesUsers"],
  ].forEach((fn) =>
    fn.addToRolePolicy(
      new PolicyStatement({
        actions: ["execute-api:ManageConnections"],
        resources: [
          `arn:aws:execute-api:${process.env.CDK_DEFAULT_REGION}:${process.env.CDK_DEFAULT_ACCOUNT}:${webSocketApi.apiId}/*`,
        ],
      })
    )
  );

  return [webSocketApi, getStageAndLogs(scope, webSocketApi)];
};
