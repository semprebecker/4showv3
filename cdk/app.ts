import { App } from 'aws-cdk-lib';

import { SetupStack } from './setup-stack';
import { envs } from '../backend/main/config';

const app = new App();

new SetupStack(app, `4showStack${envs.stage.toUpperCase()}`, {
  description: `4show Stack ${envs.stage.toUpperCase()}`,
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
  stackName: `app-4show-stack-${envs.stage}`,
});
