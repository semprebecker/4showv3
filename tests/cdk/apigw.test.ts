import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { getApiGateway, } from '../../cdk/apigw';
import { getFunctions } from '../../cdk/lambda';
import { getCognitoPool } from '../../cdk/cognito';

describe('API Gateway', () => {
  test('returns a websocket api', () => {
    const app = new App();
    const stack = new Stack(app, 'ApiTestStack', { env: { account: '123456789', region: 'us-east-1' } });
    getApiGateway(stack, getFunctions(stack), getCognitoPool(stack));
    const cfn = Template.fromStack(stack).toJSON();
    const resources = cfn.Resources;
    const matchObject: { Parameters: Record<string, unknown>; Resources: Record<string, unknown> } = {
      Parameters: expect.any(Object),
      Resources: {},
    };
    Object.keys(resources).forEach((res) => {
      switch (resources[res].Type) {
        case 'AWS::Lambda::Function':
          matchObject.Resources[res] = {
            Properties: { Code: expect.any(Object) },
          };
          break;
        default:
          break;
      }
    });

    expect(cfn).toMatchSnapshot(matchObject);
  });
});
