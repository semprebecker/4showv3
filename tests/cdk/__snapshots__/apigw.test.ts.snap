// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Gateway returns a websocket api 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "AccessLogs8B620ECA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "Account": Object {
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "CWRole7EC94C1E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "Authorizer": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "AuthorizerType": "REQUEST",
        "AuthorizerUri": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
              Object {
                "Fn::GetAtt": Array [
                  "setupconnectionauthorizerFunctionE2E31CC5",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
        "Name": "WebSocketApiAuthorizer",
      },
      "Type": "AWS::ApiGatewayV2::Authorizer",
    },
    "CWRole7EC94C1E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DEVStage3CF34D28": Object {
      "Properties": Object {
        "AccessLogSettings": Object {
          "DestinationArn": Object {
            "Fn::GetAtt": Array [
              "AccessLogs8B620ECA",
              "Arn",
            ],
          },
          "Format": "$context.identity.sourceIp - - [$context.requestTime] \\"$context.httpMethod $context.routeKey $context.protocol\\" $context.status $context.responseLength $context.requestId",
        },
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "AutoDeploy": true,
        "DefaultRouteSettings": Object {
          "DataTraceEnabled": true,
          "LoggingLevel": "INFO",
          "ThrottlingBurstLimit": 500,
          "ThrottlingRateLimit": 1000,
        },
        "StageName": "dev",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "ExecutionLogs6FEE23C8": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/apigateway/",
              Object {
                "Ref": "WebSocketApi4showDEVD00A8EAF",
              },
              "/dev",
            ],
          ],
        },
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "GetStateModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "ContentType": "application/json",
        "Name": "GetStateModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "action": Object {
              "enum": Array [
                "getState",
              ],
            },
          },
          "required": Array [
            "action",
          ],
          "title": "GetStateSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "WebSocketApi4showDEVD00A8EAF": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "Name": "WebSocketApi4showDEV",
        "ProtocolType": "WEBSOCKET",
        "RouteSelectionExpression": "$request.body.action",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "WebSocketApi4showDEVconnectRoute8E598343": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "AuthorizationType": "CUSTOM",
        "AuthorizerId": Object {
          "Ref": "Authorizer",
        },
        "RouteKey": "$connect",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "WebSocketApi4showDEVconnectRouteOnConnectIntegrationF47787F4",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApi4showDEVconnectRouteOnConnectIntegrationF47787F4": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
              Object {
                "Fn::GetAtt": Array [
                  "setupconnectiononConnectFunction1026145B",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApi4showDEVconnectRouteOnConnectIntegrationPermissionD73F152E": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "setupconnectiononConnectFunction1026145B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:123456789:",
              Object {
                "Ref": "WebSocketApi4showDEVD00A8EAF",
              },
              "/*/*$connect",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketApi4showDEVdisconnectRouteF7E47E8F": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "RouteKey": "$disconnect",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "WebSocketApi4showDEVdisconnectRouteOnDisconnectIntegrationC465D215",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApi4showDEVdisconnectRouteOnDisconnectIntegrationC465D215": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
              Object {
                "Fn::GetAtt": Array [
                  "setupconnectiononDisconnectFunction6260FCA7",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApi4showDEVdisconnectRouteOnDisconnectIntegrationPermission1684DCDB": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "setupconnectiononDisconnectFunction6260FCA7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:123456789:",
              Object {
                "Ref": "WebSocketApi4showDEVD00A8EAF",
              },
              "/*/*$disconnect",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketApi4showDEVgetStateRouteB60A3BB2": Object {
      "DependsOn": Array [
        "Account",
        "GetStateModel",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "ModelSelectionExpression": "$request.body.action",
        "RequestModels": Object {
          "getState": "GetStateModel",
        },
        "RouteKey": "getState",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "WebSocketApi4showDEVgetStateRoutegetStateIntegrationAFCF3A6B",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApi4showDEVgetStateRoutegetStateIntegrationAFCF3A6B": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "WebSocketApi4showDEVD00A8EAF",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
              Object {
                "Fn::GetAtt": Array [
                  "setupconnectiongetStateFunction1D3A554F",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApi4showDEVgetStateRoutegetStateIntegrationPermission307899ED": Object {
      "DependsOn": Array [
        "Account",
      ],
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "setupconnectiongetStateFunction1D3A554F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:123456789:",
              Object {
                "Ref": "WebSocketApi4showDEVD00A8EAF",
              },
              "/*/*getState",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "setupconnectionauthorizerFunctionAuthZPermissionA7BA042B": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "setupconnectionauthorizerFunctionE2E31CC5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:123456789:",
              Object {
                "Ref": "WebSocketApi4showDEVD00A8EAF",
              },
              "/authorizers/",
              Object {
                "Ref": "Authorizer",
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "setupconnectionauthorizerFunctionE2E31CC5": Object {
      "DependsOn": Array [
        "setupconnectionauthorizerFunctionServiceRoleCD1EC894",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "secretsId": "devF7B884FF-75G7hosnKYhv",
            "stage": "dev",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "setupconnectionauthorizerFunctionServiceRoleCD1EC894",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "setupconnectionauthorizerFunctionServiceRoleCD1EC894": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "setupconnectiongetStateFunction1D3A554F": Object {
      "DependsOn": Array [
        "setupconnectiongetStateFunctionServiceRoleDefaultPolicyA2443520",
        "setupconnectiongetStateFunctionServiceRole7399EE77",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "secretsId": "devF7B884FF-75G7hosnKYhv",
            "stage": "dev",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "setupconnectiongetStateFunctionServiceRole7399EE77",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "setupconnectiongetStateFunctionServiceRole7399EE77": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "setupconnectiongetStateFunctionServiceRoleDefaultPolicyA2443520": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "execute-api:ManageConnections",
                "cognito-idp:*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:execute-api:undefined:undefined:",
                      Object {
                        "Ref": "WebSocketApi4showDEVD00A8EAF",
                      },
                      "/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:cognito-idp:undefined:undefined:",
                      Object {
                        "Ref": "usuariosdevA198FC9E",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "setupconnectiongetStateFunctionServiceRoleDefaultPolicyA2443520",
        "Roles": Array [
          Object {
            "Ref": "setupconnectiongetStateFunctionServiceRole7399EE77",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "setupconnectiononConnectFunction1026145B": Object {
      "DependsOn": Array [
        "setupconnectiononConnectFunctionServiceRole5F640ED8",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "secretsId": "devF7B884FF-75G7hosnKYhv",
            "stage": "dev",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "setupconnectiononConnectFunctionServiceRole5F640ED8",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "setupconnectiononConnectFunctionServiceRole5F640ED8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "setupconnectiononDisconnectFunction6260FCA7": Object {
      "DependsOn": Array [
        "setupconnectiononDisconnectFunctionServiceRoleA17A7E4A",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "secretsId": "devF7B884FF-75G7hosnKYhv",
            "stage": "dev",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "setupconnectiononDisconnectFunctionServiceRoleA17A7E4A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "setupconnectiononDisconnectFunctionServiceRoleA17A7E4A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "usuariosdevA198FC9E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_email",
              "Priority": 1,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": false,
        },
        "AliasAttributes": Array [
          "email",
        ],
        "AutoVerifiedAttributes": Array [
          "email",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": Object {
          "PasswordPolicy": Object {
            "MinimumLength": 8,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "RequireUppercase": true,
          },
        },
        "Schema": Array [
          Object {
            "Mutable": true,
            "Name": "preferred_username",
            "Required": true,
          },
          Object {
            "Mutable": true,
            "Name": "email",
            "Required": true,
          },
        ],
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolName": "USUARIOS",
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
